R/interpolate.R

```diff
@@ -118,3 +117,121 @@ sqlCommentSpec <- function(start, end, endRequired) {
 sqlQuoteSpec <- function(start, end, escape = "", doubleEscape = TRUE) {
   list(start, end, escape, doubleEscape)
 }
+
+#' @export
+#' @rdname sqlParseVariables
+#' @param sql SQL to parse (a character vector of length 1)
+#' @param quotes A list of \code{QuoteSpec} calls defining the quoting
+#'   specification.
+#' @param comments A list of \code{CommentSpec} calls defining the commenting
+#'   specification.
+sqlParseVariablesImpl <- function(sql, quotes, comments) {
+  sql_arr <- c(strsplit(as.character(sql), "", fixed = TRUE)[[1]], " ")
+
+  # characters valid in variable names
+  var_chars <- c(LETTERS, letters, 0:9, "_")
+
+  # return values
+  var_pos_start <- integer()
+  var_pos_end <- integer()
+
+  # internal helper variables
+  quote_spec_offset <- 0L
+  comment_spec_offset <- 0L
+  sql_variable_start <- 0L
+  sql_variable_end <- 0L
+
+  # prepare comments & quotes for quicker comparisions
+  for(c in seq_along(comments)) {
+    comments[[c]][[1]] <- strsplit(comments[[c]][[1]], "", fixed = TRUE)[[1]]
```

Comment:
The [[n]] access is not very intuitive, but fastest:

```
a <- list(start = "a", end = "b", endRequired = FALSE); microbenchmark::microbenchmark(a[[3]], a[["start"]], a[["end"]], a$s, a$start, a$e, a$end, times = 100000)
Unit: nanoseconds
         expr min  lq     mean median  uq     max neval cld
       a[[3]] 125 142 201.7776    153 167   14234 1e+05  a 
 a[["start"]] 136 156 215.8544    167 184   21100 1e+05  a 
   a[["end"]] 145 164 225.6624    173 191   62662 1e+05  a 
          a$s 186 206 299.5662    218 297   43130 1e+05  ab
      a$start 151 171 265.3310    183 261   22357 1e+05  ab
          a$e 188 207 371.8202    219 298 6883402 1e+05   b
        a$end 167 188 281.5594    201 278   21501 1e+05  ab
```

I wonder how much this affects overall performance (we're talking nanoseconds here). Let's look at this in a separate PR.